pkgs/build-support/replace-direct-dependencies.nix

```diff
@@ -0,0 +1,41 @@
+{ lib, runCommandLocal, nix }:
+
+# Replace some direct dependencies of drv, not recursing into the dependency tree.
+# You likely want to use replaceDependencies instead, unless you plan to implement your own recursion mechanism.
+{ drv, replacements ? [ ] }:
+let
+  inherit (lib)
+    isStorePath substring stringLength optionalString escapeShellArgs concatMap;
+in if replacements == [ ] then
+  drv
+else
+  let
+    drvName = if isStorePath drv then
+    # Reconstruct the name from the actual store path if available.
+      substring 33 (stringLength (baseNameOf drv)) (baseNameOf drv)
+    else if drv ? drvAttrs.name then
+    # Try to get the name from the derivation arguments otherwise (for floating or deferred derivations).
+      drv.drvAttrs.name + (let outputName = drv.outputName or "out";
+      in optionalString (outputName != "out") "-${outputName}")
+    else
+      throw "cannot reconstruct the derivation name from ${drv}";
+  in runCommandLocal drvName { nativeBuildInputs = [ nix.out ]; } ''
+    createRewriteScript() {
+        while [ $# -ne 0 ]; do
+            oldBasename="$(basename "$1")"
+            newBasename="$(basename "$2")"
+            shift 2
+            if [ ''${#oldBasename} -ne ''${#newBasename} ]; then
+                echo "cannot rewrite $oldBasename to $newBasename: length does not math" >&2
+                exit 1
+            fi
+            echo "s|$oldBasename|$newBasename|g" >> rewrite.sed
```

Comment:
We could do something like `${oldBasename$/|/\\|}` here to make it safe against names with pipes in them (is that even a thing?)

Reply:
Store path names cannot contain pipes: https://github.com/NixOS/nix/blob/8a8172cd2b5ef2f6dd2d9673a6379447d780ff17/src/libstore/path.cc#L26-L31